### PR TITLE
Make maximum number of parallel search requests configurable.

### LINKED ISSUE #21403: Rank Eval: Figure out what the maximum number/ size of rated requests
Problem: In it's current implementation the rank evaluation we have no checks and balances wrt. to the number rated search requests, number of rated results. We need to figure out what sensible limits look like here. 

See for instance also `max_concurrent_searches` in [msearch](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-multi-search.html) for how to limit the number of concurrent searches.

### PR DESCRIPTION
**Pull request by MaineC**
Problem: So far all rank eval requests are being executed in parallel. If there
are more than the search thread pool can handle, or if there are other search
requests executed in parallel rank eval can fail.

Solution: Make number of max_concurrent_searches configurable.

Name of configuration parameter is analogous to msearch. Default
max_concurrent_searches set to 10: Rank_eval isn't particularly time critical so
trying to avoid being more clever than probably needed here. Can set this value
through the API to a higher value anytime.

Fixes #21403

### CODE REVIEW

File: modules/rank-eval/src/main/java/org/elasticsearch/index/rankeval/RankEvalSpec.java [hunk 1]
```diff
@@ -89,6 +93,7 @@ public RankEvalSpec(StreamInput in) throws IOException {
         if (in.readBoolean()) {
             template = new Script(in);
         }
+        maxConcurrentSearches = in.readInt();
```
Comment (cbuescher) [inline modules/rank-eval/src/main/java/org/elasticsearch/index/rankeval/RankEvalSpec.java]: nit: maybe writeVInt

File: modules/rank-eval/src/main/java/org/elasticsearch/index/rankeval/RankEvalSpec.java [hunk 2]
```diff
@@ -104,6 +109,7 @@ public void writeTo(StreamOutput out) throws IOException {
         } else {
             out.writeBoolean(false);
         }
+        out.writeInt(maxConcurrentSearches);
```
Comment (cbuescher) [inline modules/rank-eval/src/main/java/org/elasticsearch/index/rankeval/RankEvalSpec.java]: nit: maybe readVInt

File: modules/rank-eval/src/main/java/org/elasticsearch/index/rankeval/TransportRankEvalAction.java [hunk 3]
```diff
@@ -118,11 +122,28 @@ protected void doExecute(RankEvalRequest request, ActionListener<RankEvalRespons
             types = ratedRequest.getTypes().toArray(types);
             templatedRequest.types(types);
 
-            client.search(templatedRequest, searchListener);
+            RequestTask task = new RequestTask(templatedRequest, searchListener);
+            taskQueue.add(task);
+        }
+
+        // Execute top n tasks, further execution is triggered in RankEvalActionListener
+        for (int i = 0; (i < qualityTask.getMaxConcurrentSearches() && (! taskQueue.isEmpty())); i++) {
```
Comment (cbuescher) [inline modules/rank-eval/src/main/java/org/elasticsearch/index/rankeval/TransportRankEvalAction.java]: Maybe its possible to simplify termination condition here, something like Math.min(ratedRequests.size(), qualityTask.getMaxConcurrentSearches())? I found the and/not statement a bit difficult to follow at first glance.

File: modules/rank-eval/src/test/java/org/elasticsearch/index/rankeval/RankEvalSpecTests.java [hunk 4]
```diff
@@ -83,6 +83,7 @@ public static void afterClass() throws Exception {
         return result;
     }
 
+    // TODO missing mutation stuff
```
Comment (cbuescher) [inline modules/rank-eval/src/test/java/org/elasticsearch/index/rankeval/RankEvalSpecTests.java]: Is this part difficult to add? I think it should be straight forward, would be nice to have the parameter in mutate as well.

**Review by cbuescher — CHANGES_REQUESTED** [review]
@MaineC thanks, looks good, left a few small comments.

**Comment by MaineC** [comment]
Updated, moved the missing test which I only captured in a todo here into a separate pr.

**Review by cbuescher — APPROVED** [review]

**Comment by cbuescher** [comment]
@MaineC thanks, LGTM

**Pull request merged**